### PR TITLE
FTP client is now able to connect to server over TLS by set `useTls =…

### PR DESCRIPTION
**Pull request by ba0f3**
… true` in newAsyncFtpClient proc

### CODE REVIEW

File: lib/pure/asyncftpclient.nim [hunk 1]
```diff
@@ -195,12 +219,37 @@ proc connect*(ftp: AsyncFtpClient) {.async.} =
...
   if ftp.pass != "":
     assertReply(await(ftp.send("PASS " & ftp.pass)), "230")
 
+  if ftp.useTls:
+    assertReply(await(ftp.send("PBSZ 0")), "200")
+    assertReply(await(ftp.send("PROT P")), "200")
+    assertReply(await(ftp.send("TYPE I")), "200")
+
+proc close*(ftp: AsyncFtpClient) {.async.} =
+  # Close connection to FTP server
```
Comment (konsumlamm) [inline lib/pure/asyncftpclient.nim]:
```suggestion
  ## Closes connection to FTP server.
```

File: lib/pure/asyncftpclient.nim [hunk 2]
```diff
@@ -429,6 +484,15 @@ proc newAsyncFtpClient*(address: string, port = Port(21),
...
+    when defined(ssl):
+      result.useTls = true
+      if sslContext == nil:
+        result.sslContext = getSSLContext()
+      else:
+        result.sslContext = sslContext
+    else:
+      {.error: "TLS support is not available. Cannot connect over TLS. Compile with -d:ssl to enable."}
 
 when not defined(testing) and isMainModule:
```
Comment (konsumlamm) [inline lib/pure/asyncftpclient.nim]: Future work: tests should be moved to `tests/stdlib/tasyncftpclient.nim`.

File: lib/pure/asyncftpclient.nim [hunk 3]
```diff
@@ -217,16 +266,21 @@ proc cdup*(ftp: AsyncFtpClient) {.async.} =
...
     else:
-      result.add(r & "\n")
-
-  assertReply(await(ftp.expectReply()), "226")
+      lines.add(r)
+  result = lines.join("\n")
+  let reply = await(ftp.expectReply())
+  echo reply
+  #assertReply(reply, "226")
+  #assertReply(await(ftp.expectReply()), "226")
```
Comment (dom96) [inline lib/pure/asyncftpclient.nim]:
```suggestion
```

File: lib/pure/asyncftpclient.nim [hunk 4]
```diff
@@ -360,26 +415,26 @@ proc doUpload(ftp: AsyncFtpClient, file: File,
   assert ftp.dsockConnected
 
   let total = file.getFileSize()
-  var data = newString(4000)
+  var data = alloc(4000)
```
Comment (dom96) [inline lib/pure/asyncftpclient.nim]: Why `alloc` here? `newString` makes more sense and is less dangerous.

**Review by dom96 — CHANGES_REQUESTED** [review]
Seems okay in general.

Two things:

* Don't use `alloc`, I've left an inline comment about this.
* I'm not too sure about the `close` logic. If you're going to be closing the `csock` and `dsock` on an SSL exception then shouldn't it also happen for non-tls? It might just be best not to close and keep the semantics as they are, or do so in a separate PR.

**Comment by ba0f3** [comment]
Got it, I will make changes soon05:51, March 2, 2021, Dominik Picheta <notifications@github.com>:
@dom96 requested changes on this pull request.

Seems okay in general.
Two things:

Don't use alloc, I've left an inline comment about this.
I'm not too sure about the close logic. If you're going to be closing the csock and dsock on an SSL exception then shouldn't it also happen for non-tls? It might just be best not to close and keep the semantics as they are, or do so in a separate PR.


In lib/pure/asyncftpclient.nim:
> +  echo reply
+  #assertReply(reply, "226")
+  #assertReply(await(ftp.expectReply()), "226")


⬇️ Suggested change
-  echo reply
-  #assertReply(reply, "226")
-  #assertReply(await(ftp.expectReply()), "226")





In lib/pure/asyncftpclient.nim:
> @@ -360,26 +415,26 @@ proc doUpload(ftp: AsyncFtpClient, file: File,
   assert ftp.dsockConnected
 
   let total = file.getFileSize()
-  var data = newString(4000)
+  var data = alloc(4000)

Why alloc here? newString makes more sense and is less dangerous.

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.
-- Sent from Yandex.Mail for mobile

File: lib/pure/asyncftpclient.nim (hunk 4 above)
Comment (ba0f3) [reply lib/pure/asyncftpclient.nim]: I changed it back to use string

File: lib/pure/asyncftpclient.nim (hunk 3 above)
Comment (ba0f3) [reply lib/pure/asyncftpclient.nim]: I removed `close` proc as suggested

File: lib/pure/asyncftpclient.nim [hunk 5]
```diff
@@ -217,14 +257,16 @@ proc cdup*(ftp: AsyncFtpClient) {.async.} =
...
   while ftp.dsockConnected:
     let r = await ftp.dsock.recvLine()
-    if r == "":
+    if r.len == 0:
+      ftp.dsock.close()
       ftp.dsockConnected = false
     else:
-      result.add(r & "\n")
+      lines.add(r)
+  result = lines.join("\n")
```
Comment (Araq) [inline lib/pure/asyncftpclient.nim]: Unnecessarily inefficient. Instead use something like:

```nim
  if result.len > 0: result.add "\n"
  result.add r
```

File: lib/pure/asyncftpclient.nim [hunk 6]
```diff
@@ -368,8 +409,7 @@ proc doUpload(ftp: AsyncFtpClient, file: File,
   while ftp.dsockConnected:
     if sendFut == nil or sendFut.finished:
       # TODO: Async file reading.
-      let len = file.readBuffer(addr(data[0]), 4000)
-      setLen(data, len)
```
Comment (dom96) [inline lib/pure/asyncftpclient.nim:375]: Why removing this? If we don't do this I believe we might read uninitialised bytes.

File: lib/pure/asyncftpclient.nim [hunk 7]
```diff
@@ -79,7 +79,17 @@
...
+from net import BufferSize, SslContext
+
+when defined(ssl):
+  from net import SslHandshakeType, newContext, SslCVerifyMode
+  var defaultSslContext {.threadvar.}: SslContext
+
+  proc getSSLContext(): SslContext =
+    if defaultSSLContext == nil:
+      defaultSSLContext = newContext(verifyMode = CVerifyPeer)
+    result = defaultSSLContext
```
Comment (dom96) [inline lib/pure/asyncftpclient.nim:91]: This is the same as httpclient afaik, would be nice to clean this up in the future so we don't have each module having its own `defaultSSLContext`. But that can happen in a follow up PR.

**Review by dom96 — APPROVED** [review]
Overall lgtm

**Comment by ringabout** [comment]
@ba0f3 Hello, would you please solve the comments?

File: lib/pure/asyncftpclient.nim [hunk 8]
```diff
@@ -368,8 +409,7 @@ proc doUpload(ftp: AsyncFtpClient, file: File,
   while ftp.dsockConnected:
     if sendFut == nil or sendFut.finished:
       # TODO: Async file reading.
-      let len = file.readBuffer(addr(data[0]), 4000)
-      setLen(data, len)
+      let len = file.readBuffer(addr data[0], 4000)
```
Comment (ringabout) [inline lib/pure/asyncftpclient.nim:415]:
```suggestion
      let len = file.readBuffer(addr data[0], 4000)
      setLen(data, len)
```

File: lib/pure/asyncftpclient.nim (hunk 6 above)
Comment (ringabout) [reply lib/pure/asyncftpclient.nim:375]: solved https://github.com/nim-lang/Nim/pull/17219/commits/7d49cc7feb7642e48ae544600fcd044ba64d2b74

**Review by ringabout — APPROVED** [review]
LGTM

**Pull request merged**